#2893 Move builders from dspace-server-webapp to dspace-api.

mwoodiupui opened the pull request:
## References
#2893

## Description
Move DSO Builders from dspace-server-webapp to dspace-api and share.

## Instructions for Reviewers

List of changes in this PR:
* Moved most Builders to dspace-api.  REST-specific code was not moved.
* **NOTE**:  most Builders moved from package `org.dspace.app.rest.builder` to `org.dspace.builder`.
* Took up some recommendations from code quality tools.
* Renamed an IT from AuthorizeConfigIntegrationTest to AuthorizeConfigIT because the old name made it run with the unit tests.
* **NOTE**:  this changes the default Maven treatment of tests from "do not build or run" (maven.test.skip) to "build but do not run" (skipTests).
* `dspace-api:org.dspace.app.bulkedit.MetadataExportTest` now uses Builder classes instead of creating DSOs "by hand" to demonstrate that the newly-common Builder classes now work in both `dspace-api` and `dspace-server-webapp`.

How to test:
Run all test suites.  They should show the same results as before.

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

mwoodiupui committed:
#2893 Move builders from dspace-server-webapp to dspace-api.

mwoodiupui committed:
#2893 Rebase on current main, patch recently-added class, tidy documentation.

mwoodiupui committed:
Fix bad LGTM default that suppresses test JARs.

mwoodiupui committed:
Add shebang to lgtm-build in attempt to make it work.

mwoodiupui committed:
Try again to make LGTM run custom command

mwoodiupui committed:
Yet another attempt to understand undocumented LGTM build command(s)

mwoodiupui committed:
Trim lgtm build command some more

mwoodiupui committed:
Set JAVA_HOME trying to make LGTM work as UNdocumented.

mwoodiupui committed:
Clean up a lot of ErrorProne and Checkstyle warnings.  #2983
So I can see what else is wrong.

mwoodiupui committed:
#2983 More Checkstyle errors

mwoodiupui committed:
Use skipTests not maven.test.skip so tests always built; fix dependencies. #2983

mwoodiupui committed:
#2893 Reconfigure test building on LGTM

mwoodiupui committed:
Merge branch 'main' into #2893

mwoodiupui committed:
#2893 try LGTM with default command again

mwoodiupui committed:
#2893 Re-enable test running in Travis; document changes to how-to-test.

mwoodiupui committed:
#2893 Discard un-needed custom LGTM build script.

mwoodiupui committed:
#2893 Rework a dspace-api IT to use Builders as a demo.

tdonohue commented on pom.xml:
```diff
@@ -475,29 +475,33 @@
+        <!-- Skip running Integration Tests by default, but allow override on
+             command line by setting property "-DskipITs=false".
+             ITs are always built, because we have modules that depend
+	     on common test infrastructure. -->
```
Again, tiny alignment issue cause of a tab on the last line

tdonohue reviewed (changes requested):
@mwoodiupui : Looks great overall!  I just had a few minor comments/questions inline (and minor cleanup/fixing to your example of using Builders in `dspace-api`).  Honestly, though this is excellent to see us be able to move Builders up to dspace-api, so thanks for your work on this!

**NOTE TO SELF**: When this gets merged, we'll need to warn the DSpace 7 team to rebase & change the package of all Builders in currently open REST PRs.

mwoodiupui commented on dspace-api/src/test/java/org/dspace/app/bulkedit/MetadataExportIT.java:
```diff
@@ -14,58 +14,60 @@
+        WorkspaceItem wi = WorkspaceItemBuilder.createWorkspaceItem(context, collection).build();
         Item item = wi.getItem();
         itemService.addMetadata(context, item, "dc", "contributor", "author", null, "Donald, Smith");
         item = installItemService.installItem(context, wi);
```
Thank you.  That's much neater.  I'm not at all familiar with *using* the Builder classes yet....

mwoodiupui commented on dspace-api/src/test/java/org/dspace/discovery/MockSolrSearchCore.java:
```diff
@@ -22,4 +22,8 @@ public void afterPropertiesSet() throws Exception {
     }
 
+    public void reset() {
+        // This method intentionally left blank.
```
I added it because AbstractIntegrationTestWithDatabase calls it.

There used to be two MockSolrSearchCore classes, one in dspace-api and one in dspace-server-webapp.  I think that when I merged them, I kept the wrong one.  I'll try it the other way.

mwoodiupui replied:
Thank you for asking.  This opened an interesting can of worms.  I hadn't copied enough from dspace-server-webapp, which incorrectly allowed this to succeed.

mwoodiupui committed:
#2893 Tabs to spaces.

mwoodiupui committed:
#2893 More proper use of Builders in demo test suite.

mwoodiupui committed:
Address questions about dummy methods.  This breaks tests. #2893
Merge more from dspace-server-webapp to make tests work again.
Tidy up XML layout.  Improve alignment of declared to actual dependencies.

mwoodiupui commented on dspace-api/src/test/java/org/dspace/statistics/MockSolrLoggerServiceImpl.java:
```diff
@@ -62,18 +64,28 @@ public void afterPropertiesSet() throws Exception {
+
+    /** Remove all records. */
+    public void reset() {
+        mockSolrServer.reset();
```
It's called from AbstractIntegrationTestWith Database.  I've tried to improve the JavaDoc.

mwoodiupui committed:
#2893 dspace/modules/additions also needs new dependencies, updates.

tdonohue commented on dspace-api/pom.xml:
```diff
@@ -537,6 +618,35 @@
+        <dependency>
+            <groupId>org.apache.lucene</groupId>
+            <artifactId>lucene-core</artifactId>
+            <version>${solr.client.version}</version>
```
Should this have a `<scope>test</scope>` as well? 

I'm not exactly sure why this new dependency was added, as it doesn't appear on `main` branch at all.  I'm assuming it's a dependency convergence error in the test environment?

mwoodiupui replied:
org.dspace.app.util.IndexVersion (in src/main) imports five classes from this artifact.  I probably fixed it up while using `mvn dependency:analyze` to track down an unrelated dependency puzzle.  It should be default-scoped.

tdonohue replied:
Ah, OK then. This seems fine then.  I was thrown off by the fact that lucene-core is in the midst of a bunch of other `test` dependencies.  At some point, we might want to put all the test dependencies together, separate from the non-test ones.  But, that can come in a followup PR.

tdonohue commented on dspace/modules/additions/pom.xml:
```diff
@@ -194,6 +203,35 @@
+        <dependency>
+            <groupId>org.apache.lucene</groupId>
+            <artifactId>lucene-core</artifactId>
+            <version>${solr.client.version}</version>
```
Same as above, should this have  `<scope>test</scope>` as well?

mwoodiupui replied:
I recall that, once I fixed the above, I had to add it here too, to get `additions` to build.

tdonohue commented on dspace-api/pom.xml:
```diff
@@ -323,6 +325,11 @@
+        <dependency>
+            <groupId>commons-cli</groupId>
+            <artifactId>commons-cli</artifactId>
+            <version>1.4</version>
```
Should this change be in the Parent POM (or maybe just removed)?  It looks like in the Parent POM (`[src]/pom.xml`) we already have commons-cli version 1.3.1 right now...So, it's unclear why we have a different version specified here.

mwoodiupui replied:
Thank you.  It should be declared here (since this project uses it) and updated in the parent (since it is already managed there).  I'll do that.

tdonohue commented on dspace-api/pom.xml:
```diff
@@ -537,6 +618,35 @@
+        <dependency>
+            <groupId>org.apache.xmlbeans</groupId>
+            <artifactId>xmlbeans</artifactId>
+            <version>2.6.0</version>
```
This new dependency is also a bit odd...we don't have this one specified on `main` branch.  What this a dependency conversion fix?

mwoodiupui replied:
PoiWordFilter imports from xmlbeans.  `mvn dependency:analyze` showed xmlbeans as "used undeclared".  Even though it seemed to work in practice without the declaration, I tend to tidy these up whenever I become aware of them.  If I don't know why it worked, then I don't know why it might stop working.

tdonohue replied:
Yes, I see that now after looking closer at that PoiWordFilter. This is correct then...it just is slightly unrelated to this PR, but that's OK to clean this up :)

tdonohue commented on dspace-api/pom.xml:
```diff
@@ -717,7 +827,7 @@
         <dependency>
             <groupId>org.xmlunit</groupId>
-            <artifactId>xmlunit-matchers</artifactId>
+            <artifactId>xmlunit-core</artifactId>
```
Same here...why the switch to -core?  This seems new as well as it's not on `main` branch.

mwoodiupui replied:
All of the classes that we actually use are in -core, not -matchers.  `mvn dependency:analyze` pointed this out.  I don't know why we were using -matchers.

tdonohue replied:
Ah, thanks for clarification! This seems correct then

tdonohue reviewed (changes requested):
@mwoodiupui : Overall, the code changes look good to me.  However, I had some questions about a number of the newly added dependencies in POMs.  It seems like there's several unexplained new dependencies here...and at least one seems like it might be in the wrong POM (should possibly be moved up to the Parent POM).

It's possible there's very good reasons for all this, but I'd appreciate it if you took a second look and/or maybe add in some comments in the POMs (if the reasons are easily explainable that way).

mwoodiupui committed:
#2893 Move several explicit dependency versions up to dependencyManagement in parent.

mwoodiupui committed:
Merge remote-tracking branch 'upstream/main' into #2893

mwoodiupui committed:
#2893 Fix new class that used old package for Builders.

mwoodiupui committed:
Merge remote-tracking branch 'upstream/main' into #2893

tdonohue reviewed (approved):
👍 Looks good to me now!  Thanks @mwoodiupui !  I've also run a quick test that the commands in the README (for running individual ITs or Unit Tests) still work, and they do.  So, this looks like a great addition.

tdonohue commented:
@mwoodiupui : I just realized I caused a few minor IT conflicts for you after merging a separate PR.  Could you do a quick rebase here & then I'll merge this PR next?

mwoodiupui committed:
Merge remote-tracking branch 'upstream/main' into #2893
